File: src/main/java/martian/minefactorial/datagen/client/MFBlockStateProvider.java
package martian.minefactorial.datagen.client;

import martian.minefactorial.Minefactorial;
import martian.minefactorial.content.registry.MFBlocks;
import martian.minefactorial.foundation.pipenet.AbstractPipeBlock;
import martian.minefactorial.foundation.pipenet.PipeState;
import net.minecraft.core.Direction;
import net.minecraft.data.PackOutput;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.DirectionalBlock;
import net.neoforged.neoforge.client.model.generators.*;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.registries.DeferredBlock;
import net.neoforged.neoforge.registries.DeferredHolder;

import static martian.minefactorial.Minefactorial.id;

public class MFBlockStateProvider extends BlockStateProvider {
	public MFBlockStateProvider(PackOutput output, ExistingFileHelper exFileHelper) {
		super(output, Minefactorial.MODID, exFileHelper);
	}

	private final ResourceLocation
			TEMPLATE_MACHINE = blockId("template_machine"),
			MACHINE_FRAME_SIDE = blockId("machine_frame_side"),
			MACHINE_FRAME_TOP = blockId("machine_frame_top"),
			MACHINE_FRAME_BOTTOM = blockId("machine_frame_bottom"),
			FLUID_OUTPUT_SIDE = blockId("fluid_output_side"),
			FLUID_OUTPUT_TOP = blockId("fluid_output_top"),
			FLUID_OUTPUT_BOTTOM = blockId("fluid_output_bottom"),
			FLUID_INPUT_SIDE = blockId("fluid_input_side"),
			FLUID_INPUT_TOP = blockId("fluid_input_top"),
			FLUID_INPUT_BOTTOM = blockId("fluid_input_bottom"),
			ITEM_OUTPUT_SIDE = blockId("item_output_side"),
			ITEM_OUTPUT_TOP = blockId("item_output_top"),
			ITEM_OUTPUT_BOTTOM = blockId("item_output_bottom"),
			GENERIC_OUTPUT_SIDE = blockId("generic_output_side"),
			GENERIC_OUTPUT_TOP = blockId("generic_output_top"),
			GENERIC_OUTPUT_BOTTOM = blockId("generic_output_bottom");

	@Override
	protected void registerStatesAndModels() {
		simpleBlockWithItem(MFBlocks.STEAM_TURBINE.get(), getMachineModel(MFBlocks.STEAM_TURBINE)
				.texture("side", blockId("machine/steam_turbine_side")));
		simpleBlockWithItem(MFBlocks.STEAM_BOILER.get(), getMachineModel(MFBlocks.STEAM_BOILER)
				.texture("side", blockId("machine/steam_boiler_side"))
				.texture("up", FLUID_OUTPUT_TOP));
		simpleBlockWithItem(MFBlocks.FOUNTAIN.get(), getMachineModel(MFBlocks.FOUNTAIN)
				.texture("up", FLUID_OUTPUT_TOP));
		simpleBlockWithItem(MFBlocks.CAPACITOR.get(), getMachineModel(MFBlocks.CAPACITOR)
				.texture("side", blockId("machine/capacitor_side")));
		simpleBlockWithItem(MFBlocks.PLASTIC_TANK.get(), getMachineModel(MFBlocks.PLASTIC_TANK)
				.texture("side", blockId("machine/tank_side")));
		simpleBlockWithItem(MFBlocks.CREATIVE_CAPACITOR.get(), getCreativeMachineModel(MFBlocks.CREATIVE_CAPACITOR)
				.texture("side", blockId("machine/creative_capacitor_side")));
		simpleBlockWithItem(MFBlocks.CREATIVE_TANK.get(), getCreativeMachineModel(MFBlocks.CREATIVE_TANK)
				.texture("side", blockId("machine/creative_tank_side")));

		horizontallyDirectionalBlockWithItem(MFBlocks.MOB_GRINDER, getMachineModel(MFBlocks.MOB_GRINDER)
				.texture("north", blockId("machine/mob_grinder_front"))
				.texture("south", ITEM_OUTPUT_SIDE));

		directionalBlockWithItem(MFBlocks.FLUID_EXTRACTOR, new ModelFile.UncheckedModelFile(blockId("fluid_extractor")));

		specialDirectionalMachine(
				MFBlocks.BREAKER,
				getMachineModel(MFBlocks.BREAKER)
						.texture("north", blockId("machine/breaker_front"))
						.texture("south", ITEM_OUTPUT_SIDE),
				getMachineModel(MFBlocks.BREAKER, "_up")
						.texture("up", blockId("machine/breaker_up_front"))
						.texture("down", ITEM_OUTPUT_BOTTOM),
				getMachineModel(MFBlocks.BREAKER, "_down")
						.texture("down", blockId("machine/breaker_down_front"))
						.texture("up", ITEM_OUTPUT_TOP)
		);
		specialDirectionalMachine(
				MFBlocks.PUMP,
				getMachineModel(MFBlocks.PUMP)
						.texture("north", FLUID_INPUT_SIDE)
						.texture("south", FLUID_OUTPUT_SIDE),
				getMachineModel(MFBlocks.PUMP, "_up")
						.texture("up", FLUID_INPUT_TOP)
						.texture("down", FLUID_OUTPUT_BOTTOM),
				getMachineModel(MFBlocks.PUMP, "_down")
						.texture("down", FLUID_INPUT_BOTTOM)
						.texture("up", FLUID_OUTPUT_TOP)
		);
		specialDirectionalMachine(
				MFBlocks.PLACER,
				getMachineModel(MFBlocks.PUMP).texture("north", GENERIC_OUTPUT_SIDE),
				getMachineModel(MFBlocks.PUMP, "_up").texture("up", GENERIC_OUTPUT_TOP),
				getMachineModel(MFBlocks.PUMP, "_down").texture("down", GENERIC_OUTPUT_BOTTOM)
		);

		pipeBlock(MFBlocks.ENERGY_PIPE);
		pipeBlock(MFBlocks.FLUID_PIPE);
	}

	// ID and model shorthands
	private BlockModelBuilder getMachineModel(String id) {
		return models().withExistingParent(id, TEMPLATE_MACHINE);
	}

	private BlockModelBuilder getCreativeMachineModel(String id) {
		return models().withExistingParent(id, blockId("template_machine_creative"));
	}

	private BlockModelBuilder getMachineModel(DeferredBlock<?> block) {
		return models().withExistingParent(block.getId().getPath(), blockId("template_machine"));
	}

	private BlockModelBuilder getCreativeMachineModel(DeferredBlock<?> block) {
		return models().withExistingParent(block.getId().getPath(), blockId("template_machine_creative"));
	}

	private BlockModelBuilder getMachineModel(DeferredBlock<?> block, String idAffix) {
		return models().withExistingParent(block.getId().getPath() + idAffix, blockId("template_machine"));
	}

	private BlockModelBuilder getCreativeMachineModel(DeferredBlock<?> block, String idAffix) {
		return models().withExistingParent(block.getId().getPath() + idAffix, blockId("template_machine_creative"));
	}

	private ResourceLocation blockId(String path) {
		return id("block/" + path);
	}

	// State and Model Helpers
	private void horizontallyDirectionalBlockWithItem(DeferredBlock<?> block, ModelFile model) {
		horizontalBlock(block.get(), model);
		simpleBlockItem(block.get(), model);
	}

	private void directionalBlockWithItem(DeferredBlock<?> block, ModelFile model) {
		directionalBlock(block.get(), model);
		simpleBlockItem(block.get(), model);
	}

	private void specialDirectionalMachine(DeferredBlock<?> block, ModelFile modelHorizontal, ModelFile modelUp, ModelFile modelDown) {
		getVariantBuilder(block.get()).forAllStates(state -> {
			Direction dir = state.getValue(DirectionalBlock.FACING);
			return ConfiguredModel.builder()
					.modelFile(switch (dir) {
						case DOWN -> modelDown;
						case UP -> modelUp;
						default -> modelHorizontal;
					})
					.rotationY(((int) dir.toYRot() + 180) % 360)
					.build();
		});
		simpleBlockItem(block.get(), modelHorizontal);
	}

	// Thank you Thepigcat for letting me use this code! :D
	// https://github.com/Thepigcat76/Buildcraft-Legacy/blob/f9de2cf42727334d31670c462e0e5af96cc553fe/src/main/java/com/thepigcat/fancy_pipes/datagen/FPBlockStateProvider.java#L50
	private void pipeBlock(DeferredHolder<Block, ? extends Block> block) {
		ResourceLocation loc = block.getId();
		MultiPartBlockStateBuilder builder = getMultipartBuilder(block.get());
		pipeConnection(builder, loc, Direction.UP, 180, 0);
		pipeConnection(builder, loc, Direction.DOWN, 0, 0);
		pipeConnection(builder, loc, Direction.NORTH, 90, 180);
		pipeConnection(builder, loc, Direction.EAST, 90, 270);
		pipeConnection(builder, loc, Direction.SOUTH, 90, 0);
		pipeConnection(builder, loc, Direction.WEST, 90, 90);
		builder.part().modelFile(pipeBaseModel(loc)).addModel().end();
	}

	private void pipeConnection(MultiPartBlockStateBuilder builder, ResourceLocation loc, Direction direction, int x, int y) {
		builder.part().modelFile(pipeConnectionModel(loc)).rotationX(x).rotationY(y).addModel()
				.condition(AbstractPipeBlock.getPropertyForDirection(direction), PipeState.CONNECTED).end()
				.part().modelFile(pipeBlockModel(loc)).rotationX(x).rotationY(y).addModel()
				.condition(AbstractPipeBlock.getPropertyForDirection(direction), PipeState.BLOCK).end();
	}

	private ModelFile pipeBaseModel(ResourceLocation loc) {
		return models().withExistingParent(loc.getPath() + "_base", id("block/pipe_base"))
				.texture("texture", id(loc.getNamespace(), "block/" + loc.getPath()));
	}

	private ModelFile pipeConnectionModel(ResourceLocation loc) {
		return models().withExistingParent(loc.getPath() + "_connection", id("block/pipe_connection"))
				.texture("texture", id(loc.getNamespace(), "block/" + loc.getPath()));
	}

	private ModelFile pipeBlockModel(ResourceLocation loc) {
		return models().withExistingParent(loc.getPath() + "_connection_block", id("block/pipe_connection_block"))
				.texture("texture", id(loc.getNamespace(), "block/" + loc.getPath() + "_block"));
	}
}

File: src/data/kotlin/martian/minefactorial/datagen/client/MFBlockStateProvider.kt
package martian.minefactorial.datagen.client

import martian.dapper.api.client.CubeModel
import martian.dapper.api.client.CubeModel.Companion.all
import martian.dapper.api.client.CubeModel.Companion.down
import martian.dapper.api.client.CubeModel.Companion.from
import martian.dapper.api.client.CubeModel.Companion.north
import martian.dapper.api.client.CubeModel.Companion.renderType
import martian.dapper.api.client.CubeModel.Companion.side
import martian.dapper.api.client.CubeModel.Companion.south
import martian.dapper.api.client.CubeModel.Companion.up
import martian.dapper.api.client.DapperBlockStateProvider
import martian.dapper.api.mcId
import martian.minefactorial.Minefactorial
import martian.minefactorial.content.block.foliage.BlockRubberWood
import martian.minefactorial.content.block.machinery.BlockMacerator
import martian.minefactorial.content.block.power.BlockSteamBoiler
import martian.minefactorial.content.block.redstone.BlockRedstoneClock
import martian.minefactorial.content.registry.MFBlocks
import martian.minefactorial.datagen.id
import martian.minefactorial.foundation.pipenet.AbstractPipeBlock
import martian.minefactorial.foundation.pipenet.PipeState
import net.minecraft.core.Direction
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.block.RotatedPillarBlock
import net.neoforged.neoforge.client.model.generators.ModelFile
import net.neoforged.neoforge.client.model.generators.MultiPartBlockStateBuilder
import net.neoforged.neoforge.data.event.GatherDataEvent
import net.neoforged.neoforge.registries.DeferredBlock

class MFBlockStateProvider(event: GatherDataEvent) : DapperBlockStateProvider(event, Minefactorial.MODID) {
	private val machineFrameSide = blockId("machine_frame_side")
	private val machineFrameTop = blockId("machine_frame_top")
	private val machineFrameBottom = blockId("machine_frame_bottom")
	private val creativeMachineFrameSide = blockId("creative_machine_frame_side")
	private val creativeMachineFrameTop = blockId("creative_machine_frame_top")
	private val creativeMachineFrameBottom = blockId("creative_machine_frame_bottom")
	private val fluidOutputSide = blockId("fluid_output_side")
	private val fluidOutputTop = blockId("fluid_output_top")
	private val fluidOutputBottom = blockId("fluid_output_bottom")
	private val fluidInputSide = blockId("fluid_input_side")
	private val fluidInputTop = blockId("fluid_input_top")
	private val fluidInputBottom = blockId("fluid_input_bottom")
	private val itemOutputSide = blockId("item_output_side")
	private val itemOutputTop = blockId("item_output_top")
	private val itemOutputBottom = blockId("item_output_bottom")
	private val genericOutputSide = blockId("generic_output_side")
	private val genericOutputTop = blockId("generic_output_top")
	private val genericOutputBottom = blockId("generic_output_bottom")

	override fun registerStatesAndModels() {
		val machineFrame = (CubeModel()
			up machineFrameTop
			down machineFrameBottom
			side machineFrameSide)
		val creativeMachineFrame = (CubeModel()
			up creativeMachineFrameTop
			down creativeMachineFrameBottom
			side creativeMachineFrameSide)

		MFBlocks.STEAM_TURBINE.addModel(machineFrame.copy()
			side machineId("steam_turbine_side")
			up machineId("steam_turbine_top"))
		MFBlocks.FOUNTAIN.addModel(machineFrame.copy() up fluidOutputTop)
		MFBlocks.CAPACITOR.addModel(machineFrame.copy() side machineId("capacitor_side"))
		MFBlocks.PLASTIC_TANK.addModel(machineFrame.copy() side machineId("tank_side"))
		MFBlocks.CREATIVE_CAPACITOR.addModel(creativeMachineFrame.copy() side machineId("creative_capacitor_side"))
		MFBlocks.CREATIVE_TANK.addModel(creativeMachineFrame.copy() side machineId("creative_tank_side"))
		MFBlocks.STORAGE_UNIT.addModel(machineFrame.copy() side machineId("storage_unit_side"))
		MFBlocks.MACHINE_FRAME.addModel(machineFrame)

		MFBlocks.MOB_GRINDER.addHorizontalDirectionalModel(machineFrame.copy()
			north machineId("mob_grinder_front")
			south itemOutputSide)
		MFBlocks.SMASHER.addHorizontalDirectionalModel(machineFrame.copy()
			north machineId("smasher_front")
			south itemOutputSide)

		MFBlocks.FLUID_EXTRACTOR.addDirectionalModel(CubeModel() from blockId("fluid_extractor"))

		MFBlocks.BREAKER.addSpecialDirectionalModel(
			machineFrame.copy() north machineId("breaker_front") south itemOutputSide,
			machineFrame.copy() up machineId("breaker_up_front") down itemOutputBottom,
			machineFrame.copy() down machineId("breaker_down_front") up itemOutputTop
		)
		MFBlocks.PUMP.addSpecialDirectionalModel(
			machineFrame.copy() north fluidInputSide south fluidOutputSide,
			machineFrame.copy() up fluidInputTop down fluidOutputBottom,
			machineFrame.copy() down fluidInputBottom up fluidOutputTop
		)
		MFBlocks.PLACER.addSpecialDirectionalModel(
			machineFrame.copy() north genericOutputSide,
			machineFrame.copy() up genericOutputTop,
			machineFrame.copy() down genericOutputBottom
		)
		MFBlocks.EJECTOR.addSpecialDirectionalModel(
			machineFrame.copy() north itemOutputSide,
			machineFrame.copy() up itemOutputTop,
			machineFrame.copy() down itemOutputBottom,
		)

		dapperToggleState(
			MFBlocks.REDSTONE_CLOCK.get(),
			BlockRedstoneClock.POWERED,
			CubeModel() all blockId("redstone/redstone_clock_on"),
			CubeModel() all blockId("redstone/redstone_clock_off")
		)
		dapperToggleState(
			MFBlocks.STEAM_BOILER.get(),
			BlockSteamBoiler.LIT,
			machineFrame.copy()
				side machineId("steam_boiler_side_lit")
				up fluidOutputTop,
			machineFrame.copy()
				side machineId("steam_boiler_side")
				up fluidOutputTop
		)

		getVariantBuilder(MFBlocks.MACERATOR.get()).apply {
			val modelRunning = machineFrame.copy()
				.up(machineId("macerator_top_running"))
				.north(machineId("macerator_side_running"))
				.south(itemOutputSide)
				.build("macerator_running", models())
			val modelNotRunning = machineFrame.copy()
				.up(machineId("macerator_top"))
				.north(machineId("macerator_side"))
				.south(itemOutputSide)
				.build("macerator_not_running", models())

			MFBlocks.MACERATOR.get().stateDefinition.possibleStates.forEach { state ->
				val direction = state.getValue(BlockMacerator.FACING);
				val running = state.getValue(BlockMacerator.RUNNING);

				partialState()
					.with(BlockMacerator.FACING, direction)
					.with(BlockMacerator.RUNNING, running)
					.modelForState()
					.rotationY(((direction.toYRot() + 180) % 360).toInt())
					.modelFile(if (running) modelRunning else modelNotRunning)
					.addModel()
			}

			simpleBlockItem(MFBlocks.MACERATOR.get(), modelNotRunning)
		}

		// why is the datagen for rubber wood so cursed...
		// todo: dapper-ify this
		getVariantBuilder(MFBlocks.RUBBER_WOOD.get()).apply {
			// We build these models early so that we don't build any duplicates as we're going
			val modelVertical = CubeModel()
				.side(blockId("foliage/rubber_wood"))
				.up(blockId("foliage/rubber_wood_end"))
				.down(blockId("foliage/rubber_wood_end"))
				.build("rubber_wood_vertical", models())
			val modelHorizontal = CubeModel()
				.all(blockId("foliage/rubber_wood"))
				.north(blockId("foliage/rubber_wood_end"))
				.south(blockId("foliage/rubber_wood_end"))
				.build("rubber_wood_horizontal", models())
			val modelVerticalWithRubber = CubeModel()
				.withParent(modelVertical.uncheckedLocation)
				.side(blockId("foliage/rubber_wood_with_rubber"))
				.build("rubber_wood_vertical_with_rubber", models())
			val modelHorizontalWithRubber = CubeModel()
				.all(blockId("foliage/rubber_wood_with_rubber"))
				.north(blockId("foliage/rubber_wood_end"))
				.south(blockId("foliage/rubber_wood_end"))
				.build("rubber_wood_horizontal_with_rubber", models())

			partialState()
				// Non-rubbery variants
				.with(BlockRubberWood.HAS_RUBBER, false)
				.with(RotatedPillarBlock.AXIS, Direction.Axis.Y)
				.modelForState()
				.modelFile(modelVertical)
				.addModel()
				.partialState()
				.with(BlockRubberWood.HAS_RUBBER, false)
				.with(RotatedPillarBlock.AXIS, Direction.Axis.Z)
				.modelForState()
				.modelFile(modelHorizontal)
				.rotationX(90)
				.addModel()
				.partialState()
				.with(BlockRubberWood.HAS_RUBBER, false)
				.with(RotatedPillarBlock.AXIS, Direction.Axis.X)
				.modelForState()
				.modelFile(modelHorizontal)
				.rotationX(90)
				.rotationY(90)
				.addModel()
				// Rubbery variants
				.partialState()
				.with(BlockRubberWood.HAS_RUBBER, true)
				.with(RotatedPillarBlock.AXIS, Direction.Axis.Y)
				.modelForState()
				.modelFile(modelVerticalWithRubber)
				.addModel()
				.partialState()
				.with(BlockRubberWood.HAS_RUBBER, true)
				.with(RotatedPillarBlock.AXIS, Direction.Axis.Z)
				.modelForState()
				.modelFile(modelHorizontalWithRubber)
				.rotationX(90)
				.addModel()
				.partialState()
				.with(BlockRubberWood.HAS_RUBBER, true)
				.with(RotatedPillarBlock.AXIS, Direction.Axis.X)
				.modelForState()
				.modelFile(modelHorizontalWithRubber)
				.rotationX(90)
				.rotationY(90)
				.addModel()

			simpleBlockItem(MFBlocks.RUBBER_WOOD.get(), modelVertical)
		}

		MFBlocks.RUBBER_LEAVES.addModel(CubeModel() all blockId("foliage/rubber_leaves") renderType "cutout".mcId)
		MFBlocks.RUBBER_SAPLING.addModel(CubeModel.crossModel(blockId("foliage/rubber_sapling")), makeItem = false)

		pipeBlock(MFBlocks.ENERGY_PIPE)
		pipeBlock(MFBlocks.FLUID_PIPE)

		MFBlocks.DECOR_BLOCKS.get("plastic_block").addModel(CubeModel() all blockId("decor/plastic_block"))
		MFBlocks.DECOR_BLOCKS.get("plastic_bricks").addModel(CubeModel() all blockId("decor/plastic_bricks"))
		MFBlocks.DECOR_BLOCKS.get("smooth_plastic").addModel(CubeModel() all blockId("decor/smooth_plastic"))
		MFBlocks.DECOR_BLOCKS.get("plastic_pillar") addAxisModel blockId("decor/plastic_pillar")

		MFBlocks.ROAD_BLOCKS.get("plastic_road").addModel(CubeModel() all blockId("decor/plastic_road"))
	}

	// Thank you Thepigcat for letting me use this code! :D
	// https://github.com/Thepigcat76/Buildcraft-Legacy/blob/f9de2cf42727334d31670c462e0e5af96cc553fe/src/main/java/com/thepigcat/fancy_pipes/datagen/FPBlockStateProvider.java#L50
	private fun pipeBlock(block: DeferredBlock<*>) {
		val loc = block.id
		val model = pipeBaseModel(loc)

		val builder = getMultipartBuilder(block.get())
		pipeConnection(builder, loc, Direction.UP, 180, 0)
		pipeConnection(builder, loc, Direction.DOWN, 0, 0)
		pipeConnection(builder, loc, Direction.NORTH, 90, 180)
		pipeConnection(builder, loc, Direction.EAST, 90, 270)
		pipeConnection(builder, loc, Direction.SOUTH, 90, 0)
		pipeConnection(builder, loc, Direction.WEST, 90, 90)
		builder.part().modelFile(model).addModel().end()

		simpleBlockItem(block.get(), model)
	}

	private fun pipeConnection(
		builder: MultiPartBlockStateBuilder,
		loc: ResourceLocation,
		direction: Direction,
		x: Int,
		y: Int
	) {
		builder.part().modelFile(pipeConnectionModel(loc)).rotationX(x).rotationY(y).addModel()
			.condition<PipeState>(
				AbstractPipeBlock.getPropertyForDirection(direction),
				PipeState.CONNECTED
			).end()
			.part().modelFile(pipeBlockModel(loc)).rotationX(x).rotationY(y).addModel()
			.condition<PipeState>(AbstractPipeBlock.getPropertyForDirection(direction), PipeState.BLOCK)
			.end()
	}

	private fun pipeBaseModel(loc: ResourceLocation): ModelFile {
		return models().withExistingParent("${loc.path}_base", "block/pipe_base".id)
			.texture("texture", Minefactorial.id(loc.namespace, "block/${loc.path}"))
	}

	private fun pipeConnectionModel(loc: ResourceLocation): ModelFile {
		return models().withExistingParent("${loc.path}_connection", "block/pipe_connection".id)
			.texture("texture", Minefactorial.id(loc.namespace, "block/${loc.path}"))
	}

	private fun pipeBlockModel(loc: ResourceLocation): ModelFile {
		return models().withExistingParent("${loc.path}_connection_block", "block/pipe_connection_block".id)
			.texture("texture", Minefactorial.id(loc.namespace, "block/${loc.path}_block"))
	}

	// ID Helpers
	private fun blockId(path: String) = "block/$path".id
	private fun machineId(path: String) = "block/machine/$path".id
}
